### DISCUSSION TITLE
Plugin operators would be a useful addition

### DISCUSSION THREAD

**Discussion by blackears** [godotengine/godot-proposals#6444] [3D]
I have some ideas for ways I'd like to extend the Godot editor, and feel that the options provided by the current plugin API are too limited.  I would like to add extensions to give artists a visual way to craft a one of my custom nodes.  I seem to be restricted to either choosing a gizmo, which is much to restrictive and a main window which is much to general.

I would like to give the user the ability to click and drag on my object to be able to add geometry to it or move bits of it around.  I'd also like to draw extra geometry in the viewport to act as visual guides.  I would also like to be able to add buttons to the toolbar that will run bits of script when they're clicked on.

The plugin gizmos limit you to just using their handles which have to be defined ahead of time.  You can't click to add new points or click on drag on anything other than the handles.  They seem mostly to be just another way to manipulate an objects' parameters.

Creating a main window on the other hand is much to general.  I'd have to completely reimplement Godot's 3D rendering pipeline to be able to use it.  I just need to draw a bit of geometry within the regular scene to comunicate with the user what is going on.

Blender's API has a concept called Operators that lets users implement custom tools.  They give the plugin access to all the mouse and key events that main window receives, the ability to draw arbitrary geometry in the viewport and also the ability to add UI elements.  Extending Godot's plugins to provide operators too would make it much easier to write rich tools to extends Godot's abilities.

**Comment by Zireael07** [comment]
Unity has ScriptableObjects in the same vein

**Reply by theraot** [reply]
For what I've been told, the closes Godot has to Unity's ScriptableObjects are Resources. I don't know what capabilities ScriptableObjects have in Unity, but… Perhaps you are thinking of the equivalent of inspector plugins? I'm aware that it is possible to extend Unity inspector… I don't know in what other ways - if any - Unity's UI can be extended.

**Comment by theraot** [comment]
Your custom node can run in the editor with `@tool` (`tool` in Godot 3.x), so the artist can interact with it. If you use `Engine.is_editor_hint` (`Engine.editor_hint` in Godot 3.x) you can have it do things only in the editor, such as letting the artist moving things inside of it or adding `MeshInstance`s that are not persisted (see `Node.owner`).

With an `EditorPlugin` you can draw over the viewport, see `EditorPlugin._forward_3d_draw_over_viewport` (`EditorPlugin.forward_spatial_draw_over_viewport` in Godot 3.x).

**Addendum**: I forgot to mention that you would also have to work the input. See `EditorPlugin._forward_3d_gui_input` (`EditorPlugin.forward_spatial_gui_input` in Godot 3.x).

And yes, with an `EditorPlugin` you can add options to the toolbar, see `EditorPlugin.add_control_to_container`.

I have done these things, and I still don't know how to make a custom Gizmo in Godot, I could never figure it out. I think you are looking in the wrong place.

By the way, the main screen plugin, it does not have to recreate Godot's 3D rendering pipeline. It is Godot, it can use Godot's 3D rendering pipeline. Make your 3D scene (or take one from the project), and instantiate it as a child of a `SubViewport` (`Viewport` in Godot 3.x), child of a `SubViewportContainer` (`ViewportContainer` in Godot 3.x), which is a `Control` child of your main screen plugin.

**Addendum**: Again about the input; It is Godot, so you can use `input_ray_pickable` or physics queries, and so forth. The same solutions you would use in a game to interact with 3D. If you already know how to do it for your game, you don't need to relearn this.

**Reply by blackears** [reply]
Thanks - that really answered a lot of questions for me.  For some reason, I had thought that nodes marked as `@tool` couldn't receive input events from the editor.
